File: ga/4omini_result_2024_7/result/message_queue_mini/10_message_queue.c
use std::ffi::CString;
use std::io::{self, Write};
use std::os::unix::io::AsRawFd;
use std::process::{Command, exit};
use nix::msg::MsgBuf;
use nix::msg::{MessageQueue, Message};
use nix::unistd::{fork, ForkFlag, execve, sleep, wait, ForkResult};
use std::ptr;

const MY_TYPE: i32 = 9527;

#[derive(Debug)]
struct Msgbuf {
    mtype: i32,
    mtext: [u8; 100],
    number: i32,
}

fn main() {
    let msgid = MessageQueue::new().expect("msgget failed");

    let pid = unsafe { fork().expect("fork failed") };

    match pid {
        ForkResult::Parent { child } => {
            // Parent process
            sleep(1);

            let mut buff = Msgbuf {
                mtype: MY_TYPE,
                mtext: [0; 100],
                number: 0,
            };

            // Get user input for mtext
            print!("Please enter a string you want to send: ");
            io::stdout().flush().expect("Failed to flush stdout");
            io::stdin().read_line(&mut String::from_utf8_lossy(&mut buff.mtext).to_string()).expect("Failed to read line");

            // Get user input for number
            print!("Please enter a number you want to send: ");
            io::stdout().flush().expect("Failed to flush stdout");
            let mut number_input = String::new();
            io::stdin().read_line(&mut number_input).expect("Failed to read line");
            buff.number = number_input.trim().parse::<i32>().expect("Invalid number");

            msgid.send(Message {
                mtype: buff.mtype,
                mtext: buff.mtext.to_vec(),
                ..Default::default()
            }).expect("msgsnd failed");

            // Wait for child process to terminate
            wait().expect("wait failed");
        },
        ForkResult::Child => {
            // Child process
            let mut buff = Msgbuf {
                mtype: MY_TYPE,
                mtext: [0; 100],
                number: 0,
            };

            let message: MsgBuf<[u8; 100]> = msgid.recv(MY_TYPE).expect("msgrcv failed");
            buff.mtext.copy_from_slice(&message.mtext);
            buff.number = message.number;

            // Print the received message
            println!("Child process read from msg: {}, {}", String::from_utf8_lossy(&buff.mtext), buff.number);
            exit(0);
        }
    }
}

File: ga/glm4_result_2024_6/result/message_queue_glm/10_message_queue.rs
use std::io::{self, Read};
use std::process::Command;
use std::os::unix::process::CommandExt;

fn main() {
    let mut input_text = String::new();
    println!("Please enter a string you want to send:");
    io::stdin().read_line(&mut input_text).expect("Failed to read line");
    let input_text = input_text.trim_end();

    let mut input_number = String::new();
    println!("Please enter a number you want to send:");
    io::stdin().read_line(&mut input_number).expect("Failed to read line");
    let input_number: i32 = input_number.trim_end().parse().expect("Please enter a valid number");

    let child_pid = unsafe {
        Command::new("sh")
            .args(&["-c", "echo $MSG_Q_ID $MSG_TYPE $input_text $input_number"])
            .spawn()
            .expect("Failed to spawn child process")
            .pid
    };

    println!("Parent process is waiting for child to finish");
    let status = Command::new("wait").arg(child_pid).status().expect("Failed to wait for child process");
    println!("Child process finished with status: {}", status);
}
